grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -104,30 +108,94 @@ protected MessageHandler(
-          .thenCompose(this::executeTracingQueryIfNeeded)
-          .whenComplete(
-              (response, error) -> {
-                if (error != null) {
-                  handleException(error);
-                } else {
-                  setSuccess(response);
-                }
-              });
     } catch (Throwable t) {
       handleException(t);
     }
   }
 
+  private void executeWithRetry(int retryCount) {
+    executeQuery()
+        .whenComplete(
+            (response, error) -> {
+              if (error != null) {
+                RetryDecision decision = shouldRetry(error, retryCount);
+                if (decision == RetryDecision.RETRY) {
+                  executeWithRetry(retryCount + 1);
+                } else if (decision == RetryDecision.RETHROW) {
+                  handleException(error);
+                }
+              } else {
+                setSuccess(response);
+              }
+            });
+  }
+
+  private RetryDecision shouldRetry(Throwable throwable, int retryCount) {
+    Optional<PersistenceException> cause = unwrapCause(throwable);
+    if (!cause.isPresent()) {
+      return RetryDecision.RETHROW;
+    }
+    PersistenceException pe = cause.get();
+    switch (pe.code()) {
+      case READ_TIMEOUT:
+        return retryPolicy.onReadTimeout((ReadTimeoutException) pe, retryCount);
+      case WRITE_TIMEOUT:
+        if (isIdempotent(throwable)) {
+          return retryPolicy.onWriteTimeout((WriteTimeoutException) pe, retryCount);
+        } else {
+          return RetryDecision.RETHROW;
+        }
+      default:
+        return RetryDecision.RETHROW;
+    }
+  }
+
+  private boolean isIdempotent(Throwable throwable) {
+    Optional<ExceptionWithIdempotencyInfo> exception =
+        unwrapExceptionWithIdempotencyInfo(throwable);
+    return exception.map(ExceptionWithIdempotencyInfo::isIdempotent).orElse(false);
+  }
+
+  private Optional<ExceptionWithIdempotencyInfo> unwrapExceptionWithIdempotencyInfo(
+      Throwable throwable) {
+    if (throwable instanceof CompletionException) {
+      return unwrapExceptionWithIdempotencyInfo(throwable.getCause());
+    } else if (throwable instanceof ExceptionWithIdempotencyInfo) {
+      return Optional.of((ExceptionWithIdempotencyInfo) throwable);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  protected Optional<PersistenceException> unwrapCause(Throwable throwable) {
+    if (throwable instanceof CompletionException
+        || throwable instanceof ExceptionWithIdempotencyInfo) {
+      return unwrapCause(throwable.getCause());
+    } else if (throwable instanceof StatusException
+        || throwable instanceof StatusRuntimeException) {
+      return Optional.empty();
+    } else if (throwable instanceof PersistenceException) {
+      return Optional.of((PersistenceException) throwable);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  public CompletionStage<Response> executeQuery() {
```

Comment:
This can be private.
Also would you mind moving it just below `executeWithRetry`? I think that would be easier to follow since it's the first method that it invokes.

Reply:
done